### ISSUE TITLE
How to access Process Manager state on error handling

### ISSUE THREAD

**Issue by marcoonroad** [commanded/commanded#123]
Hi, man! I'm developing a banking application. I've tried to figure out how to access the Process Manager/Saga's state during error handling. There's a context parameter, but I don't know if I could read my saga state within that argument. So, an `error/5` callback passing the Process Manager will be very helpful for me (and possibly for other fellows there).

**Comment by marcoonroad** [comment]
If there's a way, could you please document that into the related wiki page for Process Manager, please?

**Comment by marcoonroad** [comment]
Without that Process Manager state on error handling, we have to workaround passing a bunch of state between non-related events and commands.

**Comment by slashdotdash** [comment]
@marcoonroad Yes, I can see that access to the process manager state would be useful. 

A non-breaking change would be to provide the process manager state within the context map:

```elixir
defmodule ExampleProcessManager do
  use Commanded.ProcessManagers.ProcessManager,
    name: "ExampleProcessManager",
    router: ExampleRouter

  def error({:error, _failure}, _failed_command, _pending_commands, %{process_manager: state} = context) do
    # ... use process manager `state` to determine error handling response
  end
end
```

**Comment by marcoonroad** [comment]
:clap: :clap: :clap: :clap:

Perfect, man!

**Comment by bamorim** [comment]
@slashdotdash I'm working on this right now.

I found a way to make it backwards compatible and also add another param to the error callback.

The idea is to change the default callbacks to:
```elixir
      @doc false
      def error(error, failed_command, pending_commands, context, _process_state) do
        error(error, failed_command, pending_commands, context)
      end

      @doc false
      def error({:error, reason}, _failed_command, _pending_commands, _context),
        do: {:stop, reason}
```

And then change the ProcessManagerInstance to call the error/5. Later we can deprecate the error/4 in favor of this.
My point is that mixing the context with the process manager instance is misleading (since normally we think that the context is something defined by the users on their own error handling strategies)

What do you think?

**Comment by bamorim** [comment]
Also, to be honest, I don't see the point of having a error context while we could only have the process manager state being passed to errors and to normal handling. What was the rationale behind that decision?

**Comment by slashdotdash** [comment]
The `context` map allows you to provide state between command retry failures. The example used in the [Process Manager guide](https://github.com/commanded/commanded/blob/master/guides/Process%20Managers.md#error-handling-example) counts the number of failures and stops after too many. Without the context map there would be no way to do this.

Adding the process manager state as an additional arg and deprecating `error/4` is fine.

**Comment by bamorim** [comment]
Actually, just found a problem. Should I pass the previous, the next or both states?
(Currently, the next state is only calculated in absence of failure)

In the tests, I'm using the StartProcess and I'm saving the reply_to from ProcessStarted in the process manager state. But the problem is that ProcessStarted triggers the AttemptProcess command which is the one that fails and I don't have the "next state" in hand to get the reply_to.

**Comment by slashdotdash** [comment]
Closed by #124.
